### DISCUSSION TITLE
Can i manually connect devices??

### DISCUSSION THREAD

**Discussion by stormkuter** [DeviceFarmer/stf#739] [Q&A]
Description:
Sometimes the device connection is disconnected from the provider. It is normally connected when checked with adb devices, but it cannot 
be connected unless the provider docker is re-start. is there any way I can do it whenever I want to provide device connection?

Steps to reproduce:
I used 3.6.4 version.
connect the device to the provider pc and leave it for a long time. (about 3~4 week)

Expected result:
The device must be connected to the provider.

Debugging(error log):

{"log":"2023-12-12T03:34:37.547Z INF/device:plugins:screen:stream 21 [000009d64e966153] Launching screen service minicap-apk\n","stream":"stderr","time":"2023-12-12T03:34:37.547808961Z"}
{"log":"2023-12-12T03:34:37.547Z INF/device:resources:minicap 21 [000009d64e966153] CLASSPATH=/data/local/tmp/minicap.apk app_process /system/bin io.devicefarmer.minicap.Main -S -Q 80 -P 1080x2400@864x865/0\n","stream":"stderr","time":"2023-12-12T03:34:37.547842233Z"}
{"log":"2023-12-12T03:34:37.938Z INF/device:plugins:screen:stream 21 [000009d64e966153] minicap says: \"Failed to instantiate [ch.qos.logback.classic.LoggerContext]\"\n","stream":"stderr","time":"2023-12-12T03:34:37.938583253Z"}
{"log":"2023-12-12T03:34:37.938Z INF/device:plugins:screen:stream 21 [000009d64e966153] minicap says: \"Reported exception:\"\n","stream":"stderr","time":"2023-12-12T03:34:37.938610467Z"}
{"log":"2023-12-12T03:34:37.938Z INF/device:plugins:screen:stream 21 [000009d64e966153] minicap says: \"java.lang.ArrayIndexOutOfBoundsException: length=0; index=0\"\n","stream":"stderr","time":"2023-12-12T03:34:37.938615504Z"}
{"log":"2023-12-12T03:34:37.938Z INF/device:plugins:screen:stream 21 [000009d64e966153] minicap says: \"\u0009at android.os.Environment.getExternalStorageState(Environment.java:1279)\"\n","stream":"stderr","time":"2023-12-12T03:34:37.9386195Z"}
{"log":"2023-12-12T03:34:37.938Z INF/device:plugins:screen:stream 21 [000009d64e966153] minicap says: \"\u0009at ch.qos.logback.core.android.AndroidContextUtil.getMountedExternalStorageDirectoryPath(Unknown Source:0)\"\n","stream":"stderr","time":"2023-12-12T03:34:37.938655107Z"}
{"log":"2023-12-12T03:34:37.938Z INF/device:plugins:screen:stream 21 [000009d64e966153] minicap says: \"\u0009at ch.qos.logback.core.android.AndroidContextUtil.setupProperties(Unknown Source:14)\"\n","stream":"stderr","time":"2023-12-12T03:34:37.938672754Z"}
{"log":"2023-12-12T03:34:37.938Z INF/device:plugins:screen:stream 21 [000009d64e966153] minicap says: \"\u0009at ch.qos.logback.classic.util.ContextInitializer.autoConfig(Unknown Source:12)\"\n","stream":"stderr","time":"2023-12-12T03:34:37.938676864Z"}
{"log":"2023-12-12T03:34:37.938Z INF/device:plugins:screen:stream 21 [000009d64e966153] minicap says: \"\u0009at org.slf4j.impl.StaticLoggerBinder.init(Unknown Source:7)\"\n","stream":"stderr","time":"2023-12-12T03:34:37.938679595Z"}
{"log":"2023-12-12T03:34:37.938Z INF/device:plugins:screen:stream 21 [000009d64e966153] minicap says: \"\u0009at org.slf4j.impl.StaticLoggerBinder.\u003cclinit\u003e(Unknown Source:20)\"\n","stream":"stderr","time":"2023-12-12T03:34:37.938835281Z"}
{"log":"2023-12-12T03:34:37.938Z INF/device:plugins:screen:stream 21 [000009d64e966153] minicap says: \"\u0009at org.slf4j.LoggerFactory.bind(LoggerFactory.java:150)\"\n","stream":"stderr","time":"2023-12-12T03:34:37.93885294Z"}
{"log":"2023-12-12T03:34:37.938Z INF/device:plugins:screen:stream 21 [000009d64e966153] minicap says: \"\u0009at org.slf4j.LoggerFactory.performInitialization(LoggerFactory.java:124)\"\n","stream":"stderr","time":"2023-12-12T03:34:37.938858567Z"}
{"log":"2023-12-12T03:34:37.938Z INF/device:plugins:screen:stream 21 [000009d64e966153] minicap says: \"\u0009at org.slf4j.LoggerFactory.getILoggerFactory(LoggerFactory.java:417)\"\n","stream":"stderr","time":"2023-12-12T03:34:37.938861801Z"}
{"log":"2023-12-12T03:34:37.938Z INF/device:plugins:screen:stream 21 [000009d64e966153] minicap says: \"\u0009at org.slf4j.LoggerFactory.getLogger(LoggerFactory.java:362)\"\n","stream":"stderr","time":"2023-12-12T03:34:37.938864512Z"}
{"log":"2023-12-12T03:34:37.938Z INF/device:plugins:screen:stream 21 [000009d64e966153] minicap says: \"\u0009at io.devicefarmer.minicap.provider.BaseProvider.\u003cclinit\u003e(BaseProvider.kt:44)\"\n","stream":"stderr","time":"2023-12-12T03:34:37.938867454Z"}
{"log":"2023-12-12T03:34:37.938Z INF/device:plugins:screen:stream 21 [000009d64e966153] minicap says: \"\u0009at io.devicefarmer.minicap.Main$Companion.main(Main.kt:62)\"\n","stream":"stderr","time":"2023-12-12T03:34:37.93887042Z"}
{"log":"2023-12-12T03:34:37.938Z INF/device:plugins:screen:stream 21 [000009d64e966153] minicap says: \"\u0009at io.devicefarmer.minicap.Main.main(Unknown Source:2)\"\n","stream":"stderr","time":"2023-12-12T03:34:37.938873142Z"}
{"log":"2023-12-12T03:34:37.938Z INF/device:plugins:screen:stream 21 [000009d64e966153] minicap says: \"\u0009at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)\"\n","stream":"stderr","time":"2023-12-12T03:34:37.938875751Z"}
{"log":"2023-12-12T03:34:37.938Z INF/device:plugins:screen:stream 21 [000009d64e966153] minicap says: \"\u0009at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:359)\"\n","stream":"stderr","time":"2023-12-12T03:34:37.938878435Z"}
{"log":"2023-12-12T03:34:37.940Z INF/device:plugins:screen:stream 21 [000009d64e966153] minicap says: \"PID: 11413\"\n","stream":"stderr","time":"2023-12-12T03:34:37.940548302Z"}
{"log":"2023-12-12T03:34:37.940Z INF/device:plugins:screen:stream 21 [000009d64e966153] minicap says: \"INFO: 1080x2400@389x865/0\"\n","stream":"stderr","time":"2023-12-12T03:34:37.940561892Z"}
{"log":"2023-12-12T03:34:37.940Z INF/device:plugins:screen:stream 21 [000009d64e966153] Connecting to minicap service\n","stream":"stderr","time":"2023-12-12T03:34:37.940567432Z"}
{"log":"2023-12-12T03:34:37.942Z INF/device:plugins:screen:stream 21 [000009d64e966153] Reading minicap banner\n","stream":"stderr","time":"2023-12-12T03:34:37.942405326Z"}
...
{"log":"2023-12-12T03:35:29.688Z WRN/provider 1 [*] Device worker \"000009d64e966153\" was killed with signal SIGKILL, assuming deliberate action and not restarting\n","stream":"stderr","time":"2023-12-12T03:35:29.688342642Z"}
{"log":"2023-12-12T03:35:29.688Z INF/provider 1 [*] Cleaning up device worker \"000009d64e966153\"\n","stream":"stderr","time":"2023-12-12T03:35:29.688367013Z"}
{"log":"2023-12-12T03:35:29.688Z INF/provider 1 [*] Device worker \"000009d64e966153\" has retired\n","stream":"stderr","time":"2023-12-12T03:35:29.688394489Z"}

**Comment by denis99999** [comment]
@stormkuter , please post your issue filling the corresponding template, look also at #662

**Reply by stormkuter** [reply]
sorry, I did update it again.

**Reply by stormkuter** [reply]
@denis99999 Please check it.

**Comment by denis99999** [comment]
@stormkuter , first you should upgrade to the latest version of DeviceFarmer/STF, second I saw an error of minicap "failed to instantiate.." but it is difficult to determine the cause from the logs you provided, may be it is because a hardware connection issue or a permission issue on your phone, please provide briefly what are your architecture and configuartion: are you using USB hubs to connect phones to computers, how phones (and how many) are connected on computers, how STF service units are launched (local mode or not), operating system of computers, VM or not, etc.

**Reply by stormkuter** [reply]
okay, if i need to update, i will update to latest.
and i also will update my environments
Before that, the following error occurred and the device seems to have been disconnected, so please check.

{"log":"2023-12-14T18:29:21.561Z FTL/util:lifecycle 1 [*] Tracker had an error Error: Connection closed\n","stream":"stderr","time":"2023-12-14T18:29:21.566051076Z"}
{"log":"    at /app/node_modules/@devicefarmer/adbkit/lib/adb/tracker.js:21:13\n","stream":"stderr","time":"2023-12-14T18:29:21.566053236Z"}
{"log":"    at tryCatcher (/app/node_modules/@devicefarmer/adbkit/node_modules/bluebird/js/main/util.js:26:23)\n","stream":"stderr","time":"2023-12-14T18:29:21.566054999Z"}
{"log":"    at CatchFilter.doFilter (/app/node_modules/@devicefarmer/adbkit/node_modules/bluebird/js/main/catch_filter.js:40:36)\n","stream":"stderr","time":"2023-12-14T18:29:21.566056712Z"}
{"log":"    at CatchFilter.tryCatcher (/app/node_modules/@devicefarmer/adbkit/node_modules/bluebird/js/main/util.js:26:23)\n","stream":"stderr","time":"2023-12-14T18:29:21.566058452Z"}
{"log":"    at Promise._settlePromiseFromHandler (/app/node_modules/@devicefarmer/adbkit/node_modules/bluebird/js/main/promise.js:503:31)\n","stream":"stderr","time":"2023-12-14T18:29:21.566060291Z"}
{"log":"    at Promise._settlePromiseAt (/app/node_modules/@devicefarmer/adbkit/node_modules/bluebird/js/main/promise.js:577:18)\n","stream":"stderr","time":"2023-12-14T18:29:21.56606233Z"}
{"log":"    at Promise._settlePromises (/app/node_modules/@devicefarmer/adbkit/node_modules/bluebird/js/main/promise.js:693:14)\n","stream":"stderr","time":"2023-12-14T18:29:21.56606413Z"}
{"log":"    at Async._drainQueue (/app/node_modules/@devicefarmer/adbkit/node_modules/bluebird/js/main/async.js:123:16)\n","stream":"stderr","time":"2023-12-14T18:29:21.566065884Z"}
{"log":"    at Async._drainQueues (/app/node_modules/@devicefarmer/adbkit/node_modules/bluebird/js/main/async.js:133:10)\n","stream":"stderr","time":"2023-12-14T18:29:21.566067573Z"}
{"log":"    at Immediate.Async.drainQueues [as _onImmediate] (/app/node_modules/@devicefarmer/adbkit/node_modules/bluebird/js/main/async.js:15:14)\n","stream":"stderr","time":"2023-12-14T18:29:21.566069335Z"}
{"log":"    at processImmediate (node:internal/timers:466:21)\n","stream":"stderr","time":"2023-12-14T18:29:21.566071087Z"}

**Comment by denis99999** [comment]
@stormkuter , sorry but from information you provided I have no other idea on what cause your issue, I just know a clean deployment of STF (cf. DEPLOYMENT.md documention) is very stable in time, without untimely disconnections.

**Reply by stormkuter** [reply]
That's a shame. I drew a rough picture of my current environment. Is there anything that could be a problem?

**Reply by stormkuter** [reply]
I can't upload the image in my company.

**Reply by stormkuter** [reply]
Provider's PC is using (HP Z1 TWR Workstation G9) model.
USB Hub: Sipolar-A423
USB Cable: [VERBATIM] Cellular Phone Cable

The ADB runs in server mode on the local PC.
Provider execution command is as follows.
stf provider --name ${STATION_NAME} --connect-sub tcp://${APP_SERVER_IP}:7250 --connect-push tcp://${APP_SERVER_IP}:7270 --storage-url https://${APP_SERVER_IP}:9443/ --public-ip ${PUBLIC_IP} --heartbeat-interval 10000 --screen-ws-url-pattern "wss://${APP_SERVER_IP}:9443/d/${STATION_NAME}/<%= serial %>/<%= publicPort %>/" --adb-host host.docker.internal --min-port 7400 --max-port 7700 --allow-remote --group-timeout

Is there a problem?

**Comment by denis99999** [comment]
@stormkuter , I see you are using a USB hub 3.0 which is not recommended with STF because instability issues (cf [FAQ](https://github.com/DeviceFarmer/stf/tree/master#faq) & [Troubleshooting](https://github.com/DeviceFarmer/stf/tree/master#troubleshooting) & [Recommended hardware](https://github.com/DeviceFarmer/stf/tree/master#recommended-hardware)), so I strongly recommend you to use a USB 2.0 chain to connect your device to your PC (Hub, cable, port), there is a good chance that your problem comes from there.
Moreover, how your `provider` and `ADB` servers are launched, using Docker or through command line ? Why the option `--adb-host host.docker.internal` ? what version of `ADB` ? What operating system at `provider` side ?
Finally, as it seems you have a clean distributed architecture, separing the device providers from the main STF server, I also strongly recommend you to follow the network tuning described in #662.
